File: public/safemath.js
// Safe math for unsigned 256-bit integers using native BigInt
// Matches Go arc/safemath.go exactly

const MAX_U256 = (1n << 256n) - 1n;

export class SafeMathError extends Error {
  constructor(message) { super(message); this.name = 'SafeMathError'; }
}

export const ErrOverflow = () => new SafeMathError('arithmetic overflow');
export const ErrUnderflow = () => new SafeMathError('arithmetic underflow');
export const ErrDivisionByZero = () => new SafeMathError('division by zero');

// Clamp to U256 range [0, 2^256 - 1]
function toU256(v) {
  v = BigInt(v);
  if (v < 0n) throw ErrUnderflow();
  if (v > MAX_U256) throw ErrOverflow();
  return v;
}

// SafeAdd returns a + b, throws on overflow
export function safeAdd(a, b) {
  a = BigInt(a); b = BigInt(b);
  const result = a + b;
  if (result > MAX_U256) throw ErrOverflow();
  return result;
}

// SafeSub returns a - b, throws on underflow
export function safeSub(a, b) {
  a = BigInt(a); b = BigInt(b);
  if (b > a) throw ErrUnderflow();
  return a - b;
}

// SafeMul returns a * b, throws on overflow
export function safeMul(a, b) {
  a = BigInt(a); b = BigInt(b);
  const result = a * b;
  if (result > MAX_U256) throw ErrOverflow();
  return result;
}

// SafeDiv returns a / b, throws on division by zero
export function safeDiv(a, b) {
  a = BigInt(a); b = BigInt(b);
  if (b === 0n) throw ErrDivisionByZero();
  return a / b;
}

// SafeMod returns a % b, throws on division by zero
export function safeMod(a, b) {
  a = BigInt(a); b = BigInt(b);
  if (b === 0n) throw ErrDivisionByZero();
  return a % b;
}

// MulDiv computes (a * b) / c with full precision intermediate result
export function mulDiv(a, b, c) {
  a = BigInt(a); b = BigInt(b); c = BigInt(c);
  if (c === 0n) throw ErrDivisionByZero();
  return (a * b) / c;
}

// ConvertToShares computes shares = (assets * totalShares) / totalAssets
// Standard ERC-4626 conversion with rounding down
export function convertToShares(assets, totalShares, totalAssets) {
  assets = BigInt(assets);
  totalShares = BigInt(totalShares);
  totalAssets = BigInt(totalAssets);
  if (totalAssets === 0n) return assets; // First deposit: 1:1
  return mulDiv(assets, totalShares, totalAssets);
}

// ConvertToAssets computes assets = (shares * totalAssets) / totalShares
// Standard ERC-4626 conversion with rounding down
export function convertToAssets(shares, totalAssets, totalShares) {
  shares = BigInt(shares);
  totalAssets = BigInt(totalAssets);
  totalShares = BigInt(totalShares);
  if (totalShares === 0n) return shares; // No shares: 1:1
  return mulDiv(shares, totalAssets, totalShares);
}

// Min returns the smaller of a and b
export function min(a, b) {
  a = BigInt(a); b = BigInt(b);
  return a < b ? a : b;
}

// Max returns the larger of a and b
export function max(a, b) {
  a = BigInt(a); b = BigInt(b);
  return a > b ? a : b;
}

export { MAX_U256 };

File: arc/safemath.go
package arc

import (
	"errors"

	"github.com/holiman/uint256"
)

// Safe math errors
var (
	ErrOverflow      = errors.New("arithmetic overflow")
	ErrUnderflow     = errors.New("arithmetic underflow")
	ErrDivisionByZero = errors.New("division by zero")
)

// U256 is an alias for uint256.Int for convenience
type U256 = uint256.Int

// NewU256 creates a new U256 from a uint64
func NewU256(v uint64) *U256 {
	return uint256.NewInt(v)
}

// NewU256FromBig creates a U256 from a big.Int (returns nil if overflow)
func NewU256FromBig(b interface{ Bytes() []byte }) *U256 {
	z := new(U256)
	z.SetBytes(b.Bytes())
	return z
}

// SafeAdd returns a + b, or error if overflow
func SafeAdd(a, b *U256) (*U256, error) {
	result := new(U256)
	_, overflow := result.AddOverflow(a, b)
	if overflow {
		return nil, ErrOverflow
	}
	return result, nil
}

// SafeAdd64 returns a + b for uint64 values, or error if overflow
func SafeAdd64(a, b uint64) (uint64, error) {
	result := a + b
	if result < a {
		return 0, ErrOverflow
	}
	return result, nil
}

// SafeSub returns a - b, or error if underflow
func SafeSub(a, b *U256) (*U256, error) {
	result := new(U256)
	_, underflow := result.SubOverflow(a, b)
	if underflow {
		return nil, ErrUnderflow
	}
	return result, nil
}

// SafeSub64 returns a - b for uint64 values, or error if underflow
func SafeSub64(a, b uint64) (uint64, error) {
	if b > a {
		return 0, ErrUnderflow
	}
	return a - b, nil
}

// SafeMul returns a * b, or error if overflow
func SafeMul(a, b *U256) (*U256, error) {
	result := new(U256)
	_, overflow := result.MulOverflow(a, b)
	if overflow {
		return nil, ErrOverflow
	}
	return result, nil
}

// SafeMul64 returns a * b for uint64 values, or error if overflow
func SafeMul64(a, b uint64) (uint64, error) {
	if a == 0 || b == 0 {
		return 0, nil
	}
	result := a * b
	if result/a != b {
		return 0, ErrOverflow
	}
	return result, nil
}

// SafeDiv returns a / b, or error if division by zero
func SafeDiv(a, b *U256) (*U256, error) {
	if b.IsZero() {
		return nil, ErrDivisionByZero
	}
	result := new(U256)
	result.Div(a, b)
	return result, nil
}

// SafeDiv64 returns a / b for uint64 values, or error if division by zero
func SafeDiv64(a, b uint64) (uint64, error) {
	if b == 0 {
		return 0, ErrDivisionByZero
	}
	return a / b, nil
}

// SafeMod returns a % b, or error if division by zero
func SafeMod(a, b *U256) (*U256, error) {
	if b.IsZero() {
		return nil, ErrDivisionByZero
	}
	result := new(U256)
	result.Mod(a, b)
	return result, nil
}

// MulDiv computes (a * b) / c with full precision intermediate result
// Returns error on overflow or division by zero
func MulDiv(a, b, c *U256) (*U256, error) {
	if c.IsZero() {
		return nil, ErrDivisionByZero
	}
	// Use uint256's MulDivOverflow for precision
	result := new(U256)
	result.MulDivOverflow(a, b, c)
	return result, nil
}

// MulDiv64 computes (a * b) / c for uint64 with overflow checking
func MulDiv64(a, b, c uint64) (uint64, error) {
	if c == 0 {
		return 0, ErrDivisionByZero
	}
	// Use U256 for intermediate precision
	ua := NewU256(a)
	ub := NewU256(b)
	uc := NewU256(c)

	product := new(U256)
	product.Mul(ua, ub)

	result := new(U256)
	result.Div(product, uc)

	if !result.IsUint64() {
		return 0, ErrOverflow
	}
	return result.Uint64(), nil
}

// ConvertToShares computes shares = (assets * totalShares) / totalAssets
// This is the standard ERC-4626 conversion with rounding down
func ConvertToShares(assets, totalShares, totalAssets *U256) (*U256, error) {
	if totalAssets.IsZero() {
		// First deposit: 1:1 ratio
		return new(U256).Set(assets), nil
	}
	return MulDiv(assets, totalShares, totalAssets)
}

// ConvertToAssets computes assets = (shares * totalAssets) / totalShares
// This is the standard ERC-4626 conversion with rounding down
func ConvertToAssets(shares, totalAssets, totalShares *U256) (*U256, error) {
	if totalShares.IsZero() {
		// No shares: 1:1 ratio
		return new(U256).Set(shares), nil
	}
	return MulDiv(shares, totalAssets, totalShares)
}

// Min returns the smaller of a and b
func Min(a, b *U256) *U256 {
	if a.Lt(b) {
		return new(U256).Set(a)
	}
	return new(U256).Set(b)
}

// Max returns the larger of a and b
func Max(a, b *U256) *U256 {
	if a.Gt(b) {
		return new(U256).Set(a)
	}
	return new(U256).Set(b)
}

// Min64 returns the smaller of a and b
func Min64(a, b uint64) uint64 {
	if a < b {
		return a
	}
	return b
}

// Max64 returns the larger of a and b
func Max64(a, b uint64) uint64 {
	if a > b {
		return a
	}
	return b
}
